core/services/ocr2/plugins/ocr2vrf/blockhashes/fixed_blockhash_provider.go

```diff
@@ -41,59 +49,77 @@ func (b *fixedBlockhashProvider) OnchainVerifiableBlocks(
 	if err != nil {
 		return 0, nil, errors.Wrap(err, "current height")
 	}
-	fromBlock := toBlock - b.lookbackBlocks
-	var reqs []rpc.BatchElem
 
+	fromBlock := uint64(0)
+	if toBlock > b.lookbackBlocks {
+		fromBlock = toBlock - b.lookbackBlocks
+	}
+
+	var blockHeights []uint64
 	for i := fromBlock; i <= toBlock; i++ {
-		req := rpc.BatchElem{
-			Method: "eth_getBlockByNumber",
-			Args:   []interface{}{hexutil.EncodeBig(big.NewInt(int64(i))), false},
-			Result: &evmtypes.Head{},
+		blockHeights = append(blockHeights, i)
+	}
+
+	blockhashes := make([]common.Hash, len(blockHeights))
+
+	heads, err := b.lp.GetBlocks(blockHeights, pg.WithParentCtx(ctx))
+	if err != nil {
+		b.logger.Warnw("error in get blocks", "err", err)
+	}
+
+	for _, h := range heads {
+		blockhashes[h.BlockNumber-int64(fromBlock)] = h.BlockHash
+	}
+
+	// Fallback to RPC for blocks not found in log poller
+	var reqs []rpc.BatchElem
+	for i, bh := range blockhashes {
+		if bh == utils.EmptyHash {
+			req := rpc.BatchElem{
+				Method: "eth_getBlockByNumber",
+				Args:   []interface{}{hexutil.EncodeBig(big.NewInt(int64(i) + int64(fromBlock))), false},
+				Result: &evmtypes.Head{},
+			}
+			reqs = append(reqs, req)
 		}
-		reqs = append(reqs, req)
 	}
 
 	for i := 0; i < len(reqs); i += int(b.batchSize) {
 		j := i + int(b.batchSize)
 		if j > len(reqs) {
 			j = len(reqs)
 		}
-
 		err := b.client.BatchCallContext(ctx, reqs[i:j])
-
 		if err != nil {
 			return 0, nil, errors.Wrap(err, "batch call context eth_getBlockByNumber")
 		}
 	}
 
-	var blockhashes []common.Hash
 	for _, r := range reqs {
 		if r.Error != nil {
 			return 0, nil, errors.Wrap(r.Error, "error found in eth_getBlockByNumber response")
 		}
 		block, is := r.Result.(*evmtypes.Head)
+
 		if !is {
 			return 0, nil, errors.Errorf("expected result to be a %T, got %T", &evmtypes.Head{}, r.Result)
 		}
 		if block == nil {
 			return 0, nil, errors.New("invariant violation: got nil block")
 		}
-		if block.Hash == (common.Hash{}) {
+		if block.Hash == utils.EmptyHash {
 			return 0, nil, errors.New("missing block hash")
 		}
-		blockhashes = append(blockhashes, block.Hash)
+		blockhashes[block.Number-int64(fromBlock)] = block.Hash
 	}
 
 	return fromBlock, blockhashes, nil
 }
 
 func (b *fixedBlockhashProvider) CurrentHeight(ctx context.Context) (uint64, error) {
-	h, err := b.client.HeadByNumber(ctx, nil)
+	head, err := b.lp.LatestBlock(pg.WithParentCtx(ctx))
 	if err != nil {
-		return 0, errors.Wrap(err, "head by number")
-	}
-	if h.Number < 0 {
-		return 0, errors.Errorf("unexpected head number: %d", h.Number)
+		return 0, errors.Wrap(err, "latest block")
```

Comment:
nit: I'm generally against error wrapping unless you're adding something from the specific frame that you can't easily access at the root. If you just log an error once at the root cause you'll get the full stack trace and maximal context